src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,330 @@
+        /// </remarks>
+        /// <returns>A <see cref="IWebHostBuilder"/> instance.</returns>
+        protected virtual IWebHostBuilder CreateWebHostBuilder() =>
+            WebHostBuilderFactory.CreateFromTypesAssemblyEntryPoint<TEntryPoint>(Array.Empty<string>()) ??
+            throw new InvalidOperationException(Resources.FormatMissingCreateWebHostBuilderMethod(
+                nameof(IWebHostBuilder),
+                typeof(TEntryPoint).Assembly.EntryPoint.DeclaringType.FullName,
+                typeof(WebApplicationFactory<TEntryPoint>).Name,
+                nameof(CreateWebHostBuilder)));
+
+        /// <summary>
+        /// Creates the <see cref="TestServer"/> with the bootstrapped application in <paramref name="builder"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> used to
+        /// create the server.</param>
+        /// <returns>The <see cref="TestServer"/> with the bootstrapped application.</returns>
+        protected virtual TestServer CreateServer(IWebHostBuilder builder) => new TestServer(builder);
+
+        /// <summary>
+        /// Gives a fixture an opportunity to configure the application before it gets built.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> for the application.</param>
+        protected virtual void ConfigureWebHost(IWebHostBuilder builder)
+        {
+        }
+
+        private void EnsureBuilder()
+        {
+            if (_builder != null)
+            {
+                return;
+            }
+
+            _builder = CreateWebHostBuilder();
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="HttpClient"/> that automatically follows
+        /// redirects and handles cookies.
+        /// </summary>
+        /// <returns>The <see cref="HttpClient"/>.</returns>
+        public HttpClient CreateClient() =>
+            CreateClient(ClientOptions);
+
+        /// <summary>
+        /// Creates an instance of <see cref="HttpClient"/> that automatically follows
+        /// redirects and handles cookies.
+        /// </summary>
+        /// <returns>The <see cref="HttpClient"/>.</returns>
+        public HttpClient CreateClient(WebApplicationFactoryClientOptions options) =>
+            CreatePlainClient(options.BaseAddress, options.CreateHandlers());
+
+        /// <summary>
+        /// Creates a new instance of an <see cref="HttpClient"/> that can be used to
+        /// send <see cref="HttpRequestMessage"/> to the server. The base address of the <see cref="HttpClient"/>
+        /// instance will be set to <c>http://localhost</c>.
+        /// </summary>
+        /// <param name="handlers">A list of <see cref="DelegatingHandler"/> instances to set up on the
+        /// <see cref="HttpClient"/>.</param>
+        /// <returns>The <see cref="HttpClient"/>.</returns>
+        public HttpClient CreatePlainClient(params DelegatingHandler[] handlers) =>
```

Comment:
@Eilon I chose `CreatePlainClient` for the client without any opinions, I don't think there's a good name for it. (Empty, Vanilla, Basic)

Reply:
`CreateDefaultClient`? I feel like the term `Plain` isn't used often in API naming...